source/gcc/retarget_syscalls.c

```diff
@@ -55,6 +55,25 @@
 #define STDERR_CR_LF    0       /* STDERR: add CR for LF */
 #endif
 
+/* Forward prototypes. */
```

Comment:
Isn't there any compiler provided header file containing these?

Reply:
Not that I would know of - at least not the full set. Other implementations of retarget functions usually also provide this set of prototypes.

Reply:
I don't think there exists such a unified header - unfortunately.

Newlib's implementation of these stubs puts the declarations and definitions of these functions in the same file: https://github.com/mirror/newlib-cygwin/blob/master/libgloss/arm/syscalls.c#L20

However, there are numerous implementations of these stubs in Newlib. In one of them, the implementations (e.g., https://github.com/mirror/newlib-cygwin/blob/master/libgloss/libnosys/open.c) refer to a `_syslist.h`: https://github.com/mirror/newlib-cygwin/blob/master/newlib/libc/include/_syslist.h.
Maybe that header, in conjuction with the standard `stdio.h`, `signal.h`, `stdlib.h`, can be useful for us here?

Anyway, as these functions are meant to be overloaded by the user application (weak...), I think it makes sense to put the forward declarations in a separate header file.

Reply:
These function are not meant to be overloaded by the user application. At least not from the CMSIS-Compiler:CORE component point of view. Yes, they are weak and this is for the case when one would need to re-implement one or more functions due to some specific reason.
Implementation here, as is, serves as main (core) retarget implementation that redirects streams to other components like CMSIS-Compiler:STDIN/STDERR/STDOUT and CMSIS-Compiler:File Interface. All these have external headers that serve as APIs and may be used by user aplication.
Therefore I think there is no need for a separate header file as it only brings one additional file without any value to the user.

Header file that would provide forward prototypes/declarations for retargeted functions does not exist, because internally these functions exist without underscore (_syslist.h shall be used internally by Newlib for other purposes).

Reply:
> These function are not meant to be overloaded

> the case when one would need to re-implement one

I struggle to understand the reasoning here... Aren't these two quotes contradictory?


I understand from your explanation that the goal of CMSIS-Compiler:CORE is to just enable the other components of CMSIS-Compiler. As such, it is really internal and should not provide a user interface. Hence no header is necessary 👍. 

One way I would like to use CMSIS-Compiler:CORE is to simply avoid the linker warnings such as "_exit is not implemented and will always fail", without pulling any other feature, interface or implementation. Basically, I know I have these stubs in my project, but I don't want to be warned about them, because I know I don't use them.
So, I would like that you support using the CORE component alone, without it being an interface to any other component. What is your stance on this usage?

Reply:
In general are not meant to be overloaded and implementation should be fine for (hypothetically) 99% of applications. But for that 1% that require special handling for whatever reason, the attribute weak is added.

For example, in **_sbrk_r**, someone would like to have custom handling when there is an out of memory situation. He can keep using CMSIS-Compiler:CORE component and just re-implement _sbrk_r, add custom handling and solve his requirement.

You didn't give all the details but one can use CORE component standalone already today, for the purpose you described - to provide empty implementations (stubs) and to get rid of the warnings. However, it is not possible to only provide certain functions (i.e. if one would only like to provide _exit and no other function) - this is in part also due the linker requirements.

Reply:
Thanks for clarifying your view on this component.

We should consider how to overcome the limitations you point, but that belongs in a separate issue / PR. For now, I get what you're trying to do, and I approve of it.

Reply:
Sure, you can create a separate issue and we can discuss in details what is still missing.